A36-xds-for-servers.md

```diff
@@ -299,36 +299,94 @@ reference" and "release a reference" on the shared instance, or similar
+`grpc_server_add_insecure_http2_port()` API when using a wildcard port input.
+This arises from implementation details where Core currently invokes both
+`bind()` and `listen()` as part of `grpc_tcp_server_add_port()`, and as detailed
+earlier, since we would ideally not want to invoke `listen()` until we have a
+valid xDS configuration, `grpc_tcp_server_add_port()` is only invoked when the
+server is first ready to serve. (Note that any future transitions to the not
+serving state are still dealt with the `accept()`+`close()` method.) This
+behavior might change in the future by splitting `grpc_tcp_server_add_port` so
+that `bind()` and `listen()` are done separately, allowing `bind()` to be
+invoked when ports are added to the server and allowing the API to return the
+bound port for wildcard port inputs. Alternatively, Core might choose to follow
+the `accept()`+`close()` from the start.
+
+`grpc_server_config_fetcher_xds_create` takes a `notifier` arg of the type
+`grpc_server_xds_status_notifier`. The function pointer
+`on_serving_status_update` if not NULL configures the xDS server config fetcher
+to invoke when the serving status of the server changes. A status code of
+`GRPC_STATUS_OK` signifies that the server is serving, and not-serving
+otherwise. The API does not provide any guarantees around duplicate updates.
+
+### C++
+
+C++ will expose a new type `XdsServerBuilder` that mirrors the `ServerBuilder`
+API. The `XdsServerBuilder` will use the C core API described above to configure
+the server with the xDS server config fetcher. The server created after
+`BuildAndStart()` on the `XdsServerBuilder` will be xDS enabled.
+
+```C++
+class XdsServerServingStatusNotifierInterface {
+ public:
+  virtual ~XdsServerServingStatusNotifierInterface() = default;
+
+  // \a uri contains the listening target associated with the notification. Note
+  // that a single target provided to XdsServerBuilder can get resolved to
+  // multiple listening addresses.
+  // The callback is invoked each time there is an update to the serving status.
+  // The API does not provide any guarantees around duplicate updates.
+  // Status::OK signifies that the server is serving, while a non-OK status
+  // signifies that the server is not serving.
+  virtual void OnServingStatusUpdate(std::string uri, grpc::Status status) = 0;
+};
+
+class XdsServerBuilder : public ::grpc::ServerBuilder {
+ public:
+  // It is the responsibility of the application to make sure that \a notifier
+  // outlasts the life of the server. Notifications will start being made
+  // asynchronously once `BuildAndStart()` has been called. Note that it is
+  // possible for notifications to be made before `BuildAndStart()` returns.
+  void set_status_notifier(XdsServerServingStatusNotifierInterface* notifier);
+```
 
+The same consideration for bound port number not being available for wildcard
+port inputs provided via `AddListeningPort` applies to `XdsServerBuilder`.
 
 ### Wrapped Languages
```

Comment:
The second paragraph of this section can now be removed (the one starting with "C will expose an API for wrapped languages to utilize").